stan/math/laplace/laplace_marginal.hpp

```diff
@@ -0,0 +1,275 @@
+       const Eigen::VectorXd& a,
+       const Eigen::VectorXd& l_grad)
+      : vari(marginal_density),
+        phi_size_(phi.size()),
+        phi_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            phi_size_)),
+        theta_size_(theta.size()),
+        marginal_density_(
+          ChainableStack::instance().memalloc_.alloc_array<vari*>(1)) {
+      for (int i = 0; i < phi_size_; i++) phi_[i] = phi(i).vi_;
+
+      // CHECK -- is there a cleaner way of doing this?
+      marginal_density_[0] = this;
+      marginal_density_[0] = new vari(marginal_density, false);
+
+      // compute derivatives of covariance matrix with respect to phi.
+      covariance_sensitivities<K> f(x, theta_size_, covariance_function);
+      Eigen::MatrixXd diff_cov;
+      Eigen::MatrixXd covariance;
+      {
+        Eigen::VectorXd covariance_vector;
+        jacobian_fwd(f, value_of(phi), covariance_vector, diff_cov);
+        covariance = to_matrix(covariance_vector, theta_size_, theta_size_);
+      }
+
+      // Now compute the full gradient (using algorithm 5.1 of R & W)
+      // Check: is there an efficient way to solve / divide a diagonal matrix?
+      Eigen::MatrixXd Z;
+      {
+        Eigen::MatrixXd W_root_diag = W_root.asDiagonal();
+        Z = W_root_diag *
+              L.transpose().triangularView<Eigen::Upper>()
+               .solve(L.triangularView<Eigen::Lower>()
+                 .solve(W_root_diag));
+      }
+
+      Eigen::MatrixXd
+        C = mdivide_left_tri<Eigen::Lower>(L,
+                  diag_pre_multiply(W_root, covariance));
+
+      // CHECK -- should there be a minus sign here?
+      Eigen::VectorXd s2 = 0.5 * (covariance.diagonal()
+                 - (C.transpose() * C).diagonal())
+                 .cwiseProduct(diff_likelihood.third_diff(theta));
+
+      phi_adj_ = Eigen::VectorXd(phi_size_);
+      for (int j = 0; j < phi_size_; j++) {
+        Eigen::VectorXd j_col = diff_cov.col(j);
+        C = to_matrix(j_col, theta_size_, theta_size_);
+        double s1 = 0.5 * quad_form(C, a) - 0.5 * sum((Z * C).diagonal());
+        Eigen::VectorXd b = multiply(C, l_grad);
+        Eigen::VectorXd s3 = b - multiply(covariance, multiply(Z, b));
+        phi_adj_[j] = s1 + dot_product(s2, s3);
+      }
+    }
+
+    void chain() { 
+      for (int j = 0; j < phi_size_; j++)
+        phi_[j]->adj_ += marginal_density_[0]->adj_ * phi_adj_[j];
+    }
+  };
+
+  // A wrapper for the case where phi is passed as a var.
+  // Note: the intial guess is also allowed to be passed as a var.
+  template <typename T0, typename T1, typename D, typename K>
+  T1 laplace_marginal_density
+      (const Eigen::Matrix<T0, Eigen::Dynamic, 1>& theta_0,
+       const Eigen::Matrix<T1, Eigen::Dynamic, 1>& phi,
+       const std::vector<Eigen::VectorXd>& x,
+       const D& diff_likelihood,
+       const K& covariance_function,
+       double tolerance = 1e-6,
+       long int max_num_steps = 100) {
```

Comment:
I'd be v careful with template type deduction here and the previous function you defined. The non-templated one is usually preferred unless the template produces a slightly better deduction on the types.

Reply:
The non-template function doesn't construct the vari, which is expensive, and runs about twice as fast.

Reply:
That's reasonable. Mostly what I'm talking about here is making sure that the compiler selects the right function. template deduction rules for how the compiler chooses the templated vs. non-templated types can get a bit weird. May be better to have the non-template be a specialization